"Error HH502: Couldn't download compiler version list. Please check your internet connection and try again."

### Version of Hardhat

`2.26.1`

### What happened?

I keep getting the "Error HH502: Couldn't download compiler version list. Please check your internet connection and try again." error since around 4 days in GitHub CIs:
- https://github.com/pcaversaccio/hardhat-project-template-ts/actions/runs/16730490220/job/47357102124,
- https://github.com/pcaversaccio/batch-distributor/actions/runs/16730536801/job/47357256702,
- https://github.com/pcaversaccio/escrow-contract/actions/runs/16730650576/job/47357648032.

```bash
Run pnpm test:hh

> hardhat-project-template-ts@1.0.0 test:hh /home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts
> npx hardhat test

npm warn Unknown env config "verify-deps-before-run". This will stop working in the next major version of npm.
npm warn Unknown env config "_jsr-registry". This will stop working in the next major version of npm.
Downloading solc 0.8.30
Error HH502: Couldn't download compiler version list. Please check your internet connection and try again.
HardhatError: HH502: Couldn't download compiler version list. Please check your internet connection and try again.
    at /home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/solidity/compiler/downloader.ts:177:17
    at async MultiProcessMutex._executeFunctionAndReleaseMutex (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/util/multi-process-mutex.ts:77:19)
    at async CompilerDownloader.downloadCompiler (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/solidity/compiler/downloader.ts:162:5)
    at async SimpleTaskDefinition.action (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/builtin-tasks/compile.ts:588:7)
    at async Environment._runTaskDefinition (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/core/runtime-environment.ts:351:14)
    at async OverriddenTaskDefinition._action (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/@matterlabs+hardhat-zksync-solc@1.5.0_encoding@0.1.13_hardhat@2.26.1_bufferutil@4.0.9_t_d47f0601de26ce2b342554840137810c/node_modules/@matterlabs/hardhat-zksync-solc/src/index.ts:385:20)
    at async Environment._runTaskDefinition (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/core/runtime-environment.ts:351:14)
    at async Environment.run (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/core/runtime-environment.ts:184:14)
    at async SimpleTaskDefinition.action (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/builtin-tasks/compile.ts:740:36)
    at async Environment._runTaskDefinition (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/hardhat@2.26.1_bufferutil@4.0.9_ts-node@10.9.2_@types+node@24.2.0_typescript@5.9.2__typ_56de31f8b3ad7daabb82573b8dbf1de1/node_modules/hardhat/src/internal/core/runtime-environment.ts:351:14)

    Caused by: InvalidArgumentError: maxRedirections is not supported, use the redirect interceptor
        at new Request (node:internal/deps/undici/undici:2396:17)
        at Client.[dispatch] (node:internal/deps/undici/undici:8278:25)
        at Client.dispatch (node:internal/deps/undici/undici:1816:33)
        at Pool.[dispatch] (node:internal/deps/undici/undici:2074:32)
        at Pool.dispatch (node:internal/deps/undici/undici:1816:33)
        at Agent.[dispatch] (node:internal/deps/undici/undici:8721:27)
        at Agent.dispatch (node:internal/deps/undici/undici:1816:33)
        at Agent.request (/home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/undici@5.29.0/node_modules/undici/lib/api/api-request.js:169:10)
        at /home/runner/work/hardhat-project-template-ts/hardhat-project-template-ts/node_modules/.pnpm/undici@5.29.0/node_modules/undici/lib/api/api-request.js:162:15
        at new Promise (<anonymous>)

 ELIFECYCLE  Command failed with exit code 1.
Error: Process completed with exit code 1.
```

### Minimal reproduction steps

I can only repro in the GitHub CI for now.

--- a/.changeset/pink-schools-exercise.md
+++ b/.changeset/pink-schools-exercise.md
@@ -1,0 +1,6 @@
+---
+"hardhat": patch
+"@nomicfoundation/hardhat-verify": patch
+---
+
+Don't use `undici`'s global dispatcher, making Hardhat more stable across Node.js versions

--- a/packages/hardhat-core/src/internal/util/download.ts
+++ b/packages/hardhat-core/src/internal/util/download.ts
@@ -24,13 +24,13 @@ export async function download(
   timeoutMillis = 10000,
   extraHeaders: { [name: string]: string } = {}
 ) {
-  const { getGlobalDispatcher, ProxyAgent, request } = await import("undici");
+  const { Agent, ProxyAgent, request } = await import("undici");
 
   let dispatcher: Dispatcher;
   if (process.env.http_proxy !== undefined && shouldUseProxy(url)) {
     dispatcher = new ProxyAgent(process.env.http_proxy);
   } else {
-    dispatcher = getGlobalDispatcher();
+    dispatcher = new Agent();
   }
 
   const hardhatVersion = getHardhatVersion();

--- a/packages/hardhat-verify/src/internal/undici.ts
+++ b/packages/hardhat-verify/src/internal/undici.ts
@@ -29,13 +29,12 @@ export async function sendPostRequest(
 }
 
 function getDispatcher(): Undici.Dispatcher {
-  const { ProxyAgent, getGlobalDispatcher } =
-    require("undici") as typeof Undici;
+  const { ProxyAgent, Agent } = require("undici") as typeof Undici;
   if (process.env.http_proxy !== undefined) {
     return new ProxyAgent(process.env.http_proxy);
   }
 
-  return getGlobalDispatcher();
+  return new Agent();
 }
 
 export function isSuccessStatusCode(statusCode: number): boolean {

--- a/packages/hardhat-verify/src/internal/undici.ts
+++ b/packages/hardhat-verify/src/internal/undici.ts
@@ -28,13 +28,25 @@ export async function sendPostRequest(
   });
 }
 
+let mockDispatcher: Undici.Dispatcher | undefined;
+
 function getDispatcher(): Undici.Dispatcher {
+  if (mockDispatcher !== undefined) {
+    return mockDispatcher;
+  }
+
   const { ProxyAgent, Agent } = require("undici") as typeof Undici;
   if (process.env.http_proxy !== undefined) {
     return new ProxyAgent(process.env.http_proxy);
   }
 
   return new Agent();
+}
+
+export function setMockDispatcher(
+  dispatcher: Undici.Dispatcher | undefined
+): void {
+  mockDispatcher = dispatcher;
 }
 
 export function isSuccessStatusCode(statusCode: number): boolean {

--- a/packages/hardhat-verify/test/integration/mocks/etherscan.ts
+++ b/packages/hardhat-verify/test/integration/mocks/etherscan.ts
@@ -1,9 +1,5 @@
-import {
-  Dispatcher,
-  getGlobalDispatcher,
-  MockAgent,
-  setGlobalDispatcher,
-} from "undici";
+import { MockAgent } from "undici";
+import { setMockDispatcher } from "../../../src/internal/undici";
 
 const mockAgent = new MockAgent({
   keepAliveTimeout: 10,
@@ -13,17 +9,15 @@ const mockAgent = new MockAgent({
 const client = mockAgent.get("https://api-hardhat.etherscan.io");
 
 export const mockEnvironment = () => {
-  let globalDispatcher: Dispatcher;
   // enable network connections for everything but etherscan API
   mockAgent.enableNetConnect(/^(?!https:\/\/api-hardhat\.etherscan\.io)/);
 
   before(() => {
-    globalDispatcher = getGlobalDispatcher();
-    setGlobalDispatcher(mockAgent);
+    setMockDispatcher(mockAgent);
   });
 
   after(() => {
-    setGlobalDispatcher(globalDispatcher);
+    setMockDispatcher(undefined);
   });
 };
 

--- a/packages/hardhat-verify/test/integration/mocks/sourcify.ts
+++ b/packages/hardhat-verify/test/integration/mocks/sourcify.ts
@@ -1,9 +1,5 @@
-import {
-  Dispatcher,
-  getGlobalDispatcher,
-  MockAgent,
-  setGlobalDispatcher,
-} from "undici";
+import { MockAgent } from "undici";
+import { setMockDispatcher } from "../../../src/internal/undici";
 
 const mockAgent = new MockAgent({
   keepAliveTimeout: 10,
@@ -13,17 +9,15 @@ const mockAgent = new MockAgent({
 const client = mockAgent.get("https://sourcify.dev");
 
 export const mockEnvironmentSourcify = () => {
-  let globalDispatcher: Dispatcher;
   // enable network connections for everything but etherscan API
   mockAgent.enableNetConnect(/^(?!https:\/\/sourcify\.dev)/);
 
   before(() => {
-    globalDispatcher = getGlobalDispatcher();
-    setGlobalDispatcher(mockAgent);
+    setMockDispatcher(mockAgent);
   });
 
   after(() => {
-    setGlobalDispatcher(globalDispatcher);
+    setMockDispatcher(undefined);
   });
 };
 

